push-messaging-and-notifications/main.js

```diff
@@ -0,0 +1,175 @@
+'use strict';
+
+var API_KEY = 'AIzaSyBBh4ddPa96rQQNxqiq_qQj7sq1JdsNQUQ';
+
+var isPushEnabled = false;
+
+function sendToServer(subscriptionId, endpoint) {
+  // The curl command to trigger a push message straight from GCM
+  var curlCommand = 'curl --header "Authorization: key=' + API_KEY +
+    '" --header Content-Type:"application/json" ' + endpoint + 
+    ' -d "{\\"registration_ids\\":[\\"' + subscriptionId + '\\"]}"';
+
+  // Below is a curl command to test sending a push message
+  console.log(curlCommand);
+}
+
+function unsubscribe() {
+  var pushButton = document.querySelector('.js-push-button');
+  pushButton.disabled = true;
+
+  navigator.serviceWorker.ready.then(function(serviceWorkerRegistration) {
+    // To unsubscribe from push messaging, you need get the
+    // subcription object, which you can call unsubscribe() on.
+    serviceWorkerRegistration.pushManager.getSubscription().then(
+      function(pushSubscription) {
+        // Check we have a subscription to unsubscribe
+        if (!pushSubscription) {
+          // No subscription object, so set the state
+          // to allow the user to subscribe to push
+          isPushEnabled = false;
+          pushButton.disabled = false;
+          return;
+        }
+        
+        // We have a subcription, so call unsubscribe on it
+        pushSubscription.unsubscribe().then(function(successful) {
+          // TODO: Make a request to your server to remove
+          // the user from your data store so you don't attempt
+          // to send them push messages anymore
+
+          pushButton.disabled = false;
+          pushButton.textContent = 'Enable Push Messages';
+          isPushEnabled = false;
+
+          if (!successful) {
+            console.error('We were unable to unregister from push');
+            return;
+          }
+        }).catch(function(e) {
+          // We failed to unsubscribe, this can lead to
+          // an unusual state, so may be best to remove 
+          // the users data from your data store and 
+          // inform the user that you disabled push
+
+          console.log('Unsubscription error: ', e);
+          pushButton.disabled = false;
+        });
+      }).catch(function(e) {
+        console.error('Error thrown while unsubscribing from push messaging.', e);
+      });
+  });
+}
+
+
+function subscribeToPushManager() {
+  // Disable the button so it can't be changed while
+  // we process the permission request
+  var pushButton = document.querySelector('.js-push-button');
+  pushButton.disabled = true;
+
+  navigator.serviceWorker.ready.then(function(serviceWorkerRegistration) {
+    serviceWorkerRegistration.pushManager.subscribe()
+      .then(function(subscription) {
+        // The subscription was successful
+        isPushEnabled = true;
+
+        // TODO: Send the subscriptionId and endpoint to your server
```

Comment:
Isn't this being done in the `service-worker.js` code? If your point is that you can either do persist the subscription info from the context of the controlled page or the service worker, then make it clear that they're alternatives and not something that should be implemented in both places.